# Tarea 3 - Pregunta 3 - Calidad de un vino
---
# Matías Araya - 201173082-8
# Claudia Hazard - 201404523-9

# a)
Se cargan los dos dataset en un único dataframe y se agrega una columna indicando si
es vino tinto o blanco.

In [15]:
import pandas as pd
df_red = pd.read_csv("winequality-red.csv",sep=";") #Tinto
df_white = pd.read_csv("winequality-white.csv",sep=";") #Blanco
df = pd.concat([df_red,df_white], axis=0)
l=[]
for i in range(len(df_red)):
    l.append(0)
for i in range(len(df_white)):
    l.append(1)
df["tipo"]=l
print(len(df_red))
print(len(df_white))
print(max(df["quality"]))
df

1599
4898
9


fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0               7.4             0.700         0.00            1.90      0.076   
1               7.8             0.880         0.00            2.60      0.098   
2               7.8             0.760         0.04            2.30      0.092   
3              11.2             0.280         0.56            1.90      0.075   
4               7.4             0.700         0.00            1.90      0.076   
5               7.4             0.660         0.00            1.80      0.075   
6               7.9             0.600         0.06            1.60      0.069   
7               7.3             0.650         0.00            1.20      0.065   
8               7.8             0.580         0.02            2.00      0.073   
9               7.5             0.500         0.36            6.10      0.071   
10              6.7             0.580         0.08            1.80      0.097   
11              7.5             0.500         0.36            6.10      0.071   
12              5.6             0.615         0.00            1.60      0.089   
13              7.8             0.610         0.29            1.60      0.114   
14              8.9             0.620         0.18            3.80      0.176   
15              8.9             0.620         0.19            3.90      0.170   
16              8.5             0.280         0.56            1.80      0.092   
17              8.1             0.560         0.28            1.70      0.368   
18              7.4             0.590         0.08            4.40      0.086   
19              7.9             0.320         0.51            1.80      0.341   
20              8.9             0.220         0.48            1.80      0.077   
21              7.6             0.390         0.31            2.30      0.082   
22              7.9             0.430         0.21            1.60      0.106   
23              8.5             0.490         0.11            2.30      0.084   
24              6.9             0.400         0.14            2.40      0.085   
25              6.3             0.390         0.16            1.40      0.080   
26              7.6             0.410         0.24            1.80      0.080   
27              7.9             0.430         0.21            1.60      0.106   
28              7.1             0.710         0.00            1.90      0.080   
29              7.8             0.645         0.00            2.00      0.082   
...             ...               ...          ...             ...        ...   
4868            5.8             0.230         0.31            4.50      0.046   
4869            6.6             0.240         0.33           10.10      0.032   
4870            6.1             0.320         0.28            6.60      0.021   
4871            5.0             0.200         0.40            1.90      0.015   
4872            6.0             0.420         0.41           12.40      0.032   
4873            5.7             0.210         0.32            1.60      0.030   
4874            5.6             0.200         0.36            2.50      0.048   
4875            7.4             0.220         0.26            1.20      0.035   
4876            6.2             0.380         0.42            2.50      0.038   
4877            5.9             0.540         0.00            0.80      0.032   
4878            6.2             0.530         0.02            0.90      0.035   
4879            6.6             0.340         0.40            8.10      0.046   
4880            6.6             0.340         0.40            8.10      0.046   
4881            5.0             0.235         0.27           11.75      0.030   
4882            5.5             0.320         0.13            1.30      0.037   
4883            4.9             0.470         0.17            1.90      0.035   
4884            6.5             0.330         0.38            8.30      0.048   
4885            6.6             0.340

El dataset consta 6497 datos de los cuales 1599 son de vino tinto y 4898 de vino blanco. Ademas tiene 13 columnas de los cuales 8 miden cantidad de algún elemento, ademas de los grados alcoholicos, la densidad, el pH, el tipo y lo que fue catalogado como su calidad siendo 9/10 la máxima calidad recibida.

# b)
Se aborda el problema como clasificación binaria cambiando la calidad a buena cuando es mayor a 5 y mala en otro caso. Luego se crean data de entrnamiento y prueba.

In [16]:
df['good_quality'] = [1 if q>5 else 0 for q in df.quality] #Se califica como buena calidad cuando la calidad es mayor que 5.
del df["quality"]
#Se desordenan los datos.
df = df.sample(frac=1,random_state=42)
#Se crean los data train y test
import numpy as np
msk = np.random.rand(len(df)) < 0.75
df_train = df[msk]
df_test = df[~msk]
df

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
1504            7.0             0.170         0.74           12.80      0.045   
1419            7.7             0.640         0.21            2.20      0.077   
3162            6.8             0.390         0.34            7.40      0.020   
3091            6.3             0.280         0.47           11.20      0.040   
2433            7.4             0.350         0.20           13.90      0.054   
1297            7.2             0.530         0.14            2.10      0.064   
174             7.5             0.270         0.31           17.70      0.051   
3985            6.8             0.110         0.27            8.60      0.044   
561             9.0             0.440         0.49            2.40      0.078   
4347            7.1             0.230         0.30            2.60      0.034   
292             7.4             0.280         0.42           19.80      0.066   
665             7.8             0.280         0.22            1.40      0.056   
4886            6.2             0.210         0.28            5.70      0.028   
217             8.1             0.725         0.22            2.20      0.072   
230             5.2             0.480         0.04            1.60      0.054   
569             6.5             0.280         0.27            5.20      0.040   
1400            7.9             0.690         0.21            2.10      0.080   
2756            8.6             0.330         0.34           11.80      0.059   
3098            5.1             0.330         0.27            6.70      0.022   
2696            6.5             0.220         0.50           16.40      0.048   
3061            6.2             0.250         0.54            7.00      0.046   
3818            5.1             0.350         0.26            6.80      0.034   
1671            7.0             0.350         0.30            6.50      0.028   
4695            6.7             0.350         0.32            9.00      0.032   
1397            7.3             0.250         0.36           13.10      0.050   
2548            6.3             0.260         0.42            7.10      0.045   
1277            7.6             0.310         0.29           10.50      0.040   
1356            7.3             0.220         0.41           15.40      0.050   
496             7.8             0.520         0.25            1.90      0.081   
828             7.8             0.570         0.09            2.30      0.065   
...             ...               ...          ...             ...        ...   
4797            5.7             0.330         0.32            1.40      0.043   
2518            7.1             0.120         0.30            3.10      0.018   
1786            7.4             0.330         0.26            2.60      0.040   
2956            8.1             0.120         0.38            0.90      0.034   
1184            6.7             0.640         0.23            2.10      0.080   
4821            7.0             0.240         0.47            1.30      0.043   
3452            8.0             0.240         0.48            6.80      0.047   
3712            6.3             0.200         0.18           10.60      0.045   
834             7.3             0.205         0.31            1.70      0.060   
4012            6.4             0.240         0.49            5.80      0.053   
792             6.4             0.460         0.08            4.90      0.046   
769             7.9             0.720         0.01            1.90      0.076   
86              7.2             0.390         0.63           11.00      0.044   
130             8.0             0.745         0.56            2.00      0.118   
1320            7.6             0.190         0.42            1.50      0.044   
1572            6.9             0.220         0.49            7.00      0.063   
1845            7.4             0.440         0.20           11.50      0.049   
4632            6.3             0.300

# c)
Entrene un solo Arbol de Clasificación de múltiples niveles para resolver el problema. Puede variar los ´
hiper-parámetros que prefiera, recuerde que las decisiones no pueden ser basadas mirando el conjunto
de pruebas. Debido al desbalanceo que se produce en las dos clases mida la m´etrica F1-score [5] sobre
el conjunto de entrenamiento y de pruebas.

In [40]:
x_train = df_train
y_train = np.asarray(df_train.good_quality.values).ravel().tolist()
x_test = df_test
y_test = np.asarray(df_test.good_quality.values).ravel().tolist()
from sklearn.tree import DecisionTreeClassifier as Tree
model= Tree()
model.set_params(criterion='gini',splitter='best')
model.fit(x_train, y_train)

from sklearn.metrics import f1_score
y_true =  y_train
y_pred =  model.predict(x_train)
f1_score(y_true, y_pred, average='binary') # average="micro" average="macro" average='weighted'


1.0

# d)
Entrene un ensamblador de árboles de múltiples niveles, mediante la técnica de Random Forest. Varíe la
cantidad de árboles de decisión utilizados en el ensamblado (n estimators), realice un gráfico resumen
del F1-score de entrenamiento y de pruebas en función de este hiper-parámetro.

In [ ]:
from sklearn.ensemble import RandomForestClassifier
model = RandomForestClassifier(n_estimators=, max_depth=,n_jobs=-1)
from sklearn.metrics import f1_score
f1_score(y_true, y_pred)


# e)
 Entrene un ensamblador de árboles de múltiples niveles, mediante la técnica de AdaBoost. Varíe la
cantidad de árboles de decisión utilizados en el ensamblado (n estimators), realice un gráfico resumen
del F1-score de entrenamiento y de pruebas en función de este hiper-parámetro. Compare y analice con
la técnica utilizada en d).

In [ ]:
from sklearn.ensemble import AdaBoostClassifier
model = AdaBoostClassifier(base_estimator=Tree(max_depth=), n_estimators=)

# f)
Entrene alguna otra máquina de aprendizaje, elegida por usted, para resolver este problema. Elija los
hiper-parámetros que estime convenientes intentando aumentar el F1-score obtenido por los algoritmos
anteriores. Compare y analice estas 4 maneras de resolver el problema definido en b).

# g)
Defina un criterio para estimar la importancia de los distintos atributos en el ensamblado de Random Forest, implementelo sobre alguno de los ensambladores entrenados en d), haga un ranking de importancia de atributos >Es posible implementar este criterio sobre una técnica de boost como lo es AdaBoost?